### DISCUSSION TITLE
Thermal contact problems

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#20830] [Q&A General] [answered]
Hi, I am wondering what the best approach to model conduction between two meshes that are in direct contact but are not meshed together is? I have tried using the thermal contact action, but setting the gap conductivity to the material conductivity (which seems like the correct thing to do in my case) results in a model that never converges/has to use incredibly small time steps. Am I using the wrong method? In the below input file gap_conductivity values greater than about 10e-6 will not converge, which is odd when my material conductivity is 8e-3.

```
[Mesh]
    [Top_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 2
        ymax = 2
        zmax = 2
        xmin = 0
        ymin = 0
        zmin = 1
        boundary_name_prefix = 'Upper'
        boundary_id_offset = 10
    []
    [Bottom_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 2
        ymax = 2
        zmax = 1
        boundary_name_prefix = 'Lower'
    []
    [Combine]
        type = CombinerGenerator
        inputs = 'Bottom_Block Top_Block'
        positions = '0 0 0 0.12 0.12 0'
    []
    [Rename]
        type = RenameBoundaryGenerator
        input = Combine
        old_boundary = '10 11 12 13 14 15'
        new_boundary = 'Upper_back Upper_bottom Upper_right Upper_top Upper_left Upper_front'
    []
    [LowerSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = Rename
        combinatorial_geometry = 'z<1'
        block_id = 1
        block_name = 'Lower_Block'
    []
    [UpperSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = LowerSubDomain
        combinatorial_geometry = 'z>1'
        block_id = 2
        block_name = 'Upper_Block'
    []
[]

[Variables]
    [T]
        block = '1 2'
    []
[]

[ICs]
    [T_Lower]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 1
    []

    [T_Upper]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 2
    []
[]

[Kernels]
    [HeatConduction]
        type = HeatConduction
        variable = T
    []
    [HeatConductionTimeDerivative]
        type = HeatConductionTimeDerivative
        variable = T
    []
[]

[BCs]
    [Heat_Out_Base]
        type = ADDirichletBC
        value = 293.0
        variable = T
        boundary = 'Lower_back'
    []
    [Heat_In_Top]
        type = ADDirichletBC
        value = 350.0
        variable = T
        boundary = 'Upper_front'
    []
[]

[ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = T
    primary = 'Upper_back'
    secondary = 'Lower_front'
    emissivity_primary = 0
    emissivity_secondary = 0
    gap_conductivity = 1e-6
    #quadrature = True
    #gap_geometry_type = PLATE
  []
[]

[Materials]
    [./ti_material]
        type = GenericConstantMaterial
        prop_names = 'density thermal_conductivity poissons_ratio specific_heat'
        prop_values = '4e-6 8e-3 0.3 560'
        block = '1 2'
    [../]
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    automatic_scaling = True
    nl_abs_tol = 1e-3
    nl_max_its = 50
    l_max_its = 50
    num_steps = 20
    dt = 0.05
[]

[Outputs]
    [outfile]
        type = Exodus
        file_base = "ThermalGlueTest"
    []
[]
```

**Comment by GiudGiud** [comment]
Hello

`gap_geometry_type = PLATE` should be correct in this case right? Does it help?
What does the solution look like for the first few steps it converged? Is it going towards the right temperature gradient?

Guillaume

**Comment by Edward-Eth** [comment, marked as answer]
Sorry for the very delayed response, I moved onto a different problem but have now come back to this issue. It seems that gap heat transfer, when used to conduct between contacting but not merged meshes, conducts a lot faster than it ought too. For example, in input file 1, the material has a conductivity of 0.0057, and the "gap" has a conductivity of 0.0000000057, which is 10^6 times smaller than than the material. Despite this, viewing the exodus file shows that the conductivity across the "gap" occurs barely slower than conduction through the material, so despite what should be a very insulating layer between the two blocks of mesh, conductivity between them is almost as quick as conduction within. This explains why using the same value for conductivity for the gap and the part causes issues, as this would result in extremely quick conduction making the sim unstable and stop it from converging.

This seems to be specifically an issue with gaps that are zero seperation, ie in contact, as when I increase the gap to equivalent to the element size in the mesh, using the same conductance in the gap as in the material works correctly, as seen in this image:

<img width="963" alt="Screenshot 2022-05-17 at 16 30 20" src="https://user-images.githubusercontent.com/64844573/168849875-573192b1-87b6-4638-a8d5-217b5476220d.png">

This model was stable with gap conductivity = material conductivity, however this doesn't resolve the problem of what conductivity value to use to conduct between disjoint meshes that are in perfect contact. This model is Input File 2.

One thought therefore would be to artificially add a small gap into the model between the two blocks of mesh, this was attempted in Input File 3. I found that at a gap of 1/10 mesh size the simulation was again unstable. 

Basically, it seems that the issue is the tool is designed for a *gap* heat transfer, so as the gap diminishes to zero the rate of heat transfer grows and in the contact case diverges to an unsolvable fast transfer. This makes sense for a gap, the smaller the gap the faster the transfer, but then means that zero gap is unsolvable. Am I just using the wrong tool for this?

<details>
  <summary>Input File 1</summary>

```# Thermal contact between two blocks, has issues with conductivity values anywhere near the conductivity of the material itself so not sure what's going on here.

[Mesh]
    [Top_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 20
        xmin = 0
        ymin = 0
        zmin = 10
        boundary_name_prefix = 'Upper'
        boundary_id_offset = 10
    []
    [Bottom_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 10
        boundary_name_prefix = 'Lower'
    []
    [Combine]
        type = CombinerGenerator
        inputs = 'Bottom_Block Top_Block'
        positions = '0 0 0 0 0 0'
    []
    [LowerSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = Combine
        combinatorial_geometry = 'z<=10'
        block_id = 1
        block_name = 'Lower_Block'
    []
    [UpperSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = LowerSubDomain
        combinatorial_geometry = 'z>10'
        block_id = 2
        block_name = 'Upper_Block'
    []
[]

[Variables]
    [T]
        block = '1 2'
    []
[]

[ICs]
    [T_Lower]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 1
    []
    [T_Upper]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 2
    []
[]

[Kernels]
    [HeatConduction]
        type = HeatConduction
        variable = T
    []
    [HeatConductionTimeDerivative]
        type = HeatConductionTimeDerivative
        variable = T
    []
[]

[BCs]
    [Heat_Out_Base]
        type = ADDirichletBC
        value = 293.0
        variable = T
        boundary = 'Lower_back'
    []
    [Heat_In_Top]
        type = ADDirichletBC
        value = 350.0
        variable = T
        boundary = 'Upper_front'
    []
[]

[ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = T
    primary = 'Upper_back'
    secondary = 'Lower_front'
    emissivity_primary = 0
    emissivity_secondary = 0
    gap_conductivity = 0.0000000057
    gap_geometry_type = PLATE
  []
[]

[Materials]
    [./ti_material]
        type = GenericConstantMaterial
        prop_names = 'density thermal_conductivity specific_heat'
        prop_values = '0.0000045 0.0057 544'
        block = '1 2'
    [../]
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    automatic_scaling = True
    nl_max_its = 50
    l_max_its = 50
    num_steps = 40
    dt = 0.1
[]

[Outputs]
    [outfile]
        type = Exodus
        file_base = "SimpleThermalGlueTest"
        append_date = true
    []
[]
```
</details>

<details>
  <summary>Input File 2</summary>

```
# Thermal contact between two blocks, has issues with conductivity values anywhere near the conductivity of the material itself so not sure what's going on here.

[Mesh]
    [Top_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 9
        xmax = 20
        ymax = 20
        zmax = 20
        xmin = 0
        ymin = 0
        zmin = 11
        boundary_name_prefix = 'Upper'
        boundary_id_offset = 10
    []
    [Bottom_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 10
        boundary_name_prefix = 'Lower'
    []
    [Combine]
        type = CombinerGenerator
        inputs = 'Bottom_Block Top_Block'
        positions = '0 0 0 0 0 0'
    []
    [LowerSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = Combine
        combinatorial_geometry = 'z<=10'
        block_id = 1
        block_name = 'Lower_Block'
    []
    [UpperSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = LowerSubDomain
        combinatorial_geometry = 'z>10'
        block_id = 2
        block_name = 'Upper_Block'
    []
[]

[Variables]
    [T]
        block = '1 2'
    []
[]

[ICs]
    [T_Lower]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 1
    []

    [T_Upper]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 2
    []
[]

[Kernels]
    [HeatConduction]
        type = HeatConduction
        variable = T
    []
    [HeatConductionTimeDerivative]
        type = HeatConductionTimeDerivative
        variable = T
    []
[]

[BCs]
    [Heat_Out_Base]
        type = ADDirichletBC
        value = 293.0
        variable = T
        boundary = 'Lower_back'
    []
    [Heat_In_Top]
        type = ADDirichletBC
        value = 350.0
        variable = T
        boundary = 'Upper_front'
    []
[]

[ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = T
    primary = 'Upper_back'
    secondary = 'Lower_front'
    emissivity_primary = 0
    emissivity_secondary = 0
    gap_conductivity = 0.0057
    gap_geometry_type = PLATE
  []
[]

[Materials]
    [./ti_material]
        type = GenericConstantMaterial
        prop_names = 'density thermal_conductivity specific_heat'
        prop_values = '0.0000045 0.0057 544'
        block = '1 2'
    [../]
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    automatic_scaling = True
    nl_max_its = 50
    l_max_its = 50
    num_steps = 40
    dt = 0.1
[]

[Outputs]
    [outfile]
        type = Exodus
        file_base = "SimpleThermalGlueTest"
        append_date = true
    []
[]
```
</details>

<details>
  <summary>Input File 3</summary>

```
# Thermal contact between two blocks, has issues with conductivity values anywhere near the conductivity of the material itself so not sure what's going on here.

[Mesh]
    [Top_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 20
        xmin = 0
        ymin = 0
        zmin = 10
        boundary_name_prefix = 'Upper'
        boundary_id_offset = 10
    []
    [Bottom_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 10
        boundary_name_prefix = 'Lower'
    []
    [Combine]
        type = CombinerGenerator
        inputs = 'Bottom_Block Top_Block'
        positions = '0 0 0 0 0 0.1'
    []
    [LowerSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = Combine
        combinatorial_geometry = 'z<=10'
        block_id = 1
        block_name = 'Lower_Block'
    []
    [UpperSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = LowerSubDomain
        combinatorial_geometry = 'z>10'
        block_id = 2
        block_name = 'Upper_Block'
    []
[]

[Variables]
    [T]
        block = '1 2'
    []
[]

[ICs]
    [T_Lower]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 1
    []

    [T_Upper]
        type = ConstantIC
        value = 293.0
        variable = T
        block = 2
    []
[]

[Kernels]
    [HeatConduction]
        type = HeatConduction
        variable = T
    []
    [HeatConductionTimeDerivative]
        type = HeatConductionTimeDerivative
        variable = T
    []
[]

[BCs]
    [Heat_Out_Base]
        type = ADDirichletBC
        value = 293.0
        variable = T
        boundary = 'Lower_back'
    []
    [Heat_In_Top]
        type = ADDirichletBC
        value = 350.0
        variable = T
        boundary = 'Upper_front'
    []
[]

[ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = T
    primary = 'Upper_back'
    secondary = 'Lower_front'
    emissivity_primary = 0
    emissivity_secondary = 0
    gap_conductivity = 0.0057
    gap_geometry_type = PLATE
  []
[]

[Materials]
    [./ti_material]
        type = GenericConstantMaterial
        prop_names = 'density thermal_conductivity specific_heat'
        prop_values = '0.0000045 0.0057 544'
        block = '1 2'
    [../]
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    automatic_scaling = True
    nl_max_its = 50
    l_max_its = 50
    num_steps = 40
    dt = 0.1
[]

[Outputs]
    [outfile]
        type = Exodus
        file_base = "SimpleThermalGlueTest"
        append_date = true
    []
[]
```

</details>

**Reply by GiudGiud** [reply]
@Edward-Eth can you specify the gap_distance parameter manually?

the behavior you are describing is consistent with determining the gap distance automatically. If the gap is pretty much closed, then it's not really slowing down heat transfer

**Reply by Edward-Eth** [reply]
If I specify zero I'm guessing it will still break though?

**Reply by Edward-Eth** [reply]
I'm looking for something like the mesh to mesh hard contact feature in nastran:(https://help.mscsoftware.com/bundle/MSC_Nastran_2020/page/Nastran_Combined_Book/non_linear/ch09/TOC.Thermal.Contact1.xhtml) 
The temperature at each node on one surface is a combination of weighted temperatures from nodes on the opposite surface. Would this be more easily achieved with a matched value boundary condition or interface diffusion?

**Reply by Edward-Eth** [reply]
Not sure if this is relevant but what I'm modelling is actually one solid object, it's just that the mesh for two parts of it are disjoint, but what I'm trying to achieve is conduction as though they were one solid continuously meshed object. I've tried equal value boundary constraint but that seems to end up enforcing that the bottom of the top block should be the same temperature as the top of the bottom one, but never heats up the bottom one, so it effectively sets a dirichlet of the lower block's temperature on the underside of the top block.

**Reply by GiudGiud** [reply]
have you looked at the thermal contact action ?
https://mooseframework.inl.gov/source/actions/ThermalContactAction.html

**Reply by GiudGiud** [reply]
interfaces are usually for internal sidesets so I m not sure interface diffusion is the right object. Which one are you think of?

**Reply by Edward-Eth** [reply]
> have you looked at the thermal contact action ? https://mooseframework.inl.gov/source/actions/ThermalContactAction.html

I used that for setting up the gap heat transfer in the first instance. Is there some other type within thermal contact action that would be more suitable? The ThermalContactSystem page (which thermal contact action recommends you look at for examples and parameters) is empty so there's not much to go on there...

**Reply by recuero** [reply]
If you look at the parameters of this material: https://mooseframework.inl.gov/source/materials/GapConductance.html, you'll see options to define radiation and conductivity. 

Information for similar radiation formulation: https://mooseframework.inl.gov/source/userobjects/GapFluxModelRadiation.html.
For conduction, you can provide a constant gap conductivity or a function. There are some additional factors that appear in the parameter section.

**Reply by Edward-Eth** [reply]
> If you look at the parameters of this material: https://mooseframework.inl.gov/source/materials/GapConductance.html, you'll see options to define radiation and conductivity.
> 
> Information for similar radiation formulation: https://mooseframework.inl.gov/source/userobjects/GapFluxModelRadiation.html. For conduction, you can provide a constant gap conductivity or a function. There are some additional factors that appear in the parameter section.

Circling back onto this issues and still not understanding how to get this to work. Here's my input file for trying to use the GapConductance material, but nothing happens that indicates there's any heat transfer from one mesh to the other. Not sure how I'm supposed to implement the GapConductance as there's no example input file. It's labelled as a material so I've put it in the materials block, but that seems odd to me.

Input file:
<details>
  <summary>Input File 1</summary>

```
[Mesh]
    [Top_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 20
        xmin = 0
        ymin = 0
        zmin = 10
        boundary_name_prefix = 'Upper'
        boundary_id_offset = 10
    []
    [Bottom_Block]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 10
        ny = 10
        nz = 10
        xmax = 20
        ymax = 20
        zmax = 10
        boundary_name_prefix = 'Lower'
    []
    [Combine]
        type = CombinerGenerator
        inputs = 'Bottom_Block Top_Block'
        positions = '0 0 0 0 0 0'
    []
    [LowerSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = Combine
        combinatorial_geometry = 'z<=10'
        block_id = 1
        block_name = 'Lower_Block'
    []
    [UpperSubDomain]
        type = ParsedSubdomainMeshGenerator
        input = LowerSubDomain
        combinatorial_geometry = 'z>10'
        block_id = 2
        block_name = 'Upper_Block'
    []
[]

[Variables]
    [T_Upper]
        block = 2
    []
    [T_Lower]
        block = 1
    []
[]

[ICs]
    [T_Upper_Initial]
        type = ConstantIC
        value = 293.0
        variable = T_Upper
        block = 2
    []
    [T_Lower_Initial]
        type = ConstantIC
        value = 293.0
        variable = T_Lower
        block = 1
    []
[]

[Kernels]
    [HeatConduction_Upper]
        type = ADHeatConduction
        variable = T_Upper
        block = 2
    []
    [HeatConductionTimeDerivative_Upper]
        type = ADHeatConductionTimeDerivative
        variable = T_Upper
        block = 2
    []
    [HeatConduction_Lower]
        type = ADHeatConduction
        variable = T_Lower
        block = 1
    []
    [HeatConductionTimeDerivative_Lower]
        type = ADHeatConductionTimeDerivative
        variable = T_Lower
        block = 1
    []
[]

[BCs]
    [Heat_Out_Base]
        type = ADDirichletBC
        value = 293.0
        variable = T_Lower
        boundary = 'Lower_back'
    []
    [Heat_In_Top]
        type = ADDirichletBC
        value = 350.0
        variable = T_Upper
        boundary = 'Upper_front'
    []
[]

[Materials]
    [./ti_material]
        type = ADGenericConstantMaterial
        prop_names = 'density thermal_conductivity specific_heat youngs_modulus poissons_ratio'
        prop_values = '0.0000045 0.0057 544 1.16E+05 0.3'
        block = '1 2'
    [../]
    [GapTransferMaybe]
        type = GapConductance
        variable = T_Upper
        boundary = 'Upper_back'
        paired_boundary = 'Lower_front'
        gap_conductivity = 0.0057
        emissivity_primary = 0.0
        emissivity_secondary = 0.0
        gap_geometry_type = PLATE
        gap_temp = T_Lower
        quadrature = True
        warnings = True
    []
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    automatic_scaling = True
    nl_max_its = 50
    l_max_its = 200
    num_steps = 40
    dt = 0.5
[]

[Outputs]
    [outfile]
        type = Exodus
        file_base = "ThermalContactTesting/SimpleThermalGlueTest"
        append_date = true
    []
[]
```
</details>

**Reply by recuero** [reply]
I am not sure if you want to use or to implement a gap conductance model. If you want to use it, I'd just look into the tests in the MOOSE repository. For example, `large_gap_heat_transfer_test_cylinder`.i in the heat conduction module uses the thermal action to establish gap heat transfer physics in a problem: 

```[ThermalContact]
  [RPV_gap]
    type = GapHeatTransfer
    gap_geometry_type = 'CYLINDER'
    emissivity_primary = 0.8
    emissivity_secondary = 0.8
    variable = Tsolid
    primary = 'core_outer'
    secondary = 'rpv_inner'
    gap_conductivity = 0.1
    quadrature = true
    cylinder_axis_point_1 = '0 0 0'
    cylinder_axis_point_2 = '0 0 5'
  []
[]
```
The action creates the gap conductance material, or a constant gap conductance, if a constant value is provided through the action -- this is typically done for many MOOSE physics.

**Reply by GiudGiud** [reply]
Hello

This is a material because it defines new material properties:
- gap conductance
- derivatives wrt temperature of gap conductance
- gap thermal conductivity
which can then be used in the gap conductance constraint and other user objects for this. I m not sure though

@Edward-Eth the input files that show how to use these objects are in the second link @recuero shared
https://mooseframework.inl.gov/source/userobjects/GapFluxModelRadiation.html

@recuero @dschwen  would really help to have direct documentation on this object if it is usable outside of the thermal contact action. If it should not then let s disallow it

Guillaume

**Reply by Edward-Eth** [reply]
So I've made some slight progress, by using the test gap_heat_transfer_3D.i (modules/heat_conduction/tests/gap_heat_transfer_mortar/gap_heat_transfer_3D.i). With that file I have conduction between the blocks working again, but once more (as at the beginning of this discussion) when the distance between the blocks shrinks, the model stops converging. I can't seem to find any input files in the tests folders where the two meshes are in direct contact? They all have gaps between them approximately the size of an element in the mesh. If anyone knows of a test file (or any other input file) where meshes in direct contact conduct to each other that would be extremely helpful.

**Reply by Edward-Eth** [reply]
Additionally, when trying to use gap_heat_transfer_3D_mortar.i from the same folder, the radiation and conduction user objects are incorrect, they both feature the line `T = temp`, but when running the file it is clear that at some point that implementation has been changed and it should be `temperature = temp` instead. After fixing that I still can't get that test to run, as it gives me the following error after the 0th Nonlinear Residual is calculated: 
`/Users/eetheridge/mambaforge3/envs/moose/libmesh/include/metaphysicl/dynamic_std_array_wrapper.h/Users/eetheridge/mambaforge3/envs/moose/libmesh/include/metaphysicl/dynamic_std_array_wrapper.h, line 125, compiled May 16 2022 at 16:57:11
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1`

**Reply by recuero** [reply]
That test has a `min_ad_size = 100` requirement. You need to configure MOOSE with a proper size for the AD derivative container. 

For example:

`./configure --with-ad-indexing-type=global  --with-derivative-size=200`

Thanks for pointing out that those tests are not running. I'll fix that.
